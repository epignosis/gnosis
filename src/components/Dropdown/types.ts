import { ReactNode } from "react";
import { TypographyLevels } from "@theme/utils/typography";

export type DropdownItem = {
  label: string;
  value?: string;
  id?: string;
  icon?: JSX.Element | SVGAElement;
  className?: string;
  items?: DropdownItem[];
  category?: string;
};

export type DropdownProps = {
  list: DropdownItem[];
  children: ReactNode;
  placement?: PlacementOptions;
  isSearchable?: boolean;
  textSize?: TypographyLevels;
  fullWidth?: boolean;
  scrollToBottom?: boolean;
  onListItemSelect?: (item: DropdownItem) => void;
  emptyStateText?: string;
};

type PlacementOptions = "bottom-start" | "bottom-end" | "top-start" | "top-end" | "end-top";
