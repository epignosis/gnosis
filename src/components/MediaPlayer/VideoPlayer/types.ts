import { ReactPlayerProps } from "react-player";

export type PlayerProps = Omit<ReactPlayerProps, "url"> & {
  src: ReactPlayerProps["url"];
  containerAttrs?: React.HTMLAttributes<HTMLDivElement>;
};
