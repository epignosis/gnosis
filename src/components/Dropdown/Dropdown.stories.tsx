import React, { FC, useState } from "react";
import { Story } from "@storybook/react";
import Button from "../Button/Button";
import Radio from "../FormElements/RadioGroup/Radio";
import Checkbox from "../FormElements/CheckboxGroup/Checkbox";
import Dropdown from "./Dropdown";
import { DropdownItem, DropdownProps } from "./types";

const dropdownList: DropdownItem[] = [
  {
    label: "Category 1",
    items: [
      { label: "Option 1", value: "1" },
      {
        label: "Category 3",
        items: [
          { label: "Option 5", value: "5" },
          { label: "Option 6", value: "6", isDisabled: true },
        ],
      },
    ],
  },
  {
    label: "Category 2",
    items: [
      { label: "Option 3", value: "3" },
      { label: "Option 4", value: "4" },
    ],
  },
  { label: "Option 5", value: "5" },
];

export default {
  title: "components/Dropdown",
  component: Dropdown,
  argTypes: {
    placement: {
      control: {
        type: "select",
        options: ["bottom-start", "bottom-end", "top-start", "top-end", "end-top"],
      },
    },
    isSearchable: {
      control: "boolean",
    },
    fullWidth: {
      control: "boolean",
    },
    textSize: {
      control: {
        type: "select",
        options: ["xs", "sm", "md"],
      },
    },
    fixPlacement: {
      control: "boolean",
    },
  },
  args: {
    placement: "bottom-start",
    isSearchable: true,
    textSize: "sm",
    fullWidth: false,
    disabled: false,
  },
  decorators: [
    (Story: Story): JSX.Element => (
      <div
        style={{
          width: "100%",
          height: "100%",
          display: "flex",
          justifyContent: "center",
          marginTop: 200,
        }}
      >
        <Story />
      </div>
    ),
  ],
};

const Template: Story<DropdownProps> = (args) => {
  const [show, toggle] = useState(false);

  const toggleDropdown = () => {
    if (!args.disabled) {
      toggle((show) => !show);
    }
  };

  return (
    <Dropdown {...args}>
      <Button
        color="primary"
        as="div"
        className="filter-button"
        onClick={toggleDropdown}
        disabled={args.disabled}
      >
        Click (or hover) to {show ? "close" : "open"}
      </Button>
    </Dropdown>
  );
};

const extendedDropdownList: DropdownItem[] = [
  {
    label: "Category 1",
    items: [
      { label: "Option 1", value: "1" },
      {
        label: "Category 3",
        items: [
          { label: "Option 5", value: "5" },
          { label: "Option 6", value: "6" },
        ],
      },
    ],
  },
  {
    label: "Category 2",
    items: [
      { label: "Option 3", value: "3" },
      { label: "Option 4", value: "4" },
      { label: "Option 5", value: "3" },
      { label: "Option 6", value: "4" },
      { label: "Option 7", value: "3" },
      { label: "Option 8", value: "4" },
    ],
  },
  { label: "Option 5", value: "5" },
];

export const Default = Template.bind({});

Default.args = {
  list: dropdownList,
  prependContent: (
    <div style={{ marginBottom: "10px", textAlign: "center", borderBottom: "1px solid black" }}>
      Prepended Content
    </div>
  ),
};

export const withExtendedList = Template.bind({});

withExtendedList.args = {
  list: extendedDropdownList,
};

const mockList = [
  {
    id: "options",
    label: "Options",
    items: [
      {
        id: "score",
        value: "score",
        label: <Checkbox id="score" value="score" label={"Score"} name="Course score" />,
      },
      {
        id: "progress",
        value: "progress",
        label: (
          <Checkbox id="progress" value="progress" label={"Progress"} name="Course progress" />
        ),
      },
    ],
  },
  {
    id: "focus",
    value: "",
    label: "Focus",
    items: [
      {
        id: "all",
        value: "all",
        label: <Radio id="all" value="all" label={"All"} name="All" />,
      },
    ],
  },
];

export const withJsxElementLabels = Template.bind({});

withJsxElementLabels.args = {
  list: mockList,
  remainOpenOnSelect: true,
  isSearchable: false,
};

export const withHover = Template.bind({});

withHover.args = {
  list: mockList,
  hover: true,
  isSearchable: true,
};

export const withPrependContent = Template.bind({});

withPrependContent.args = {
  list: mockList,
  prependContent: (
    <div style={{ marginBottom: "10px", textAlign: "center", borderBottom: "1px solid black" }}>
      Prepended Content
    </div>
  ),
  hover: true,
  isSearchable: true,
};

const smallDropdownList: DropdownItem[] = [
  { label: "Option 1", value: "1" },
  { label: "Option 2", value: "2" },
  { label: "Option 3", value: "3" },
  { label: "Option 4", value: "4" },
];

const ScrollableContainer: FC = ({ children }) => (
  <div
    id="scrollable-container"
    style={{
      backgroundColor: "#F5F5F6",
      height: "350px",
      width: "100%",
      overflow: "auto",
      padding: "1rem",
      display: "flex",
    }}
  >
    {children}
  </div>
);

const ScrollableTemplate: Story<DropdownProps & React.CSSProperties> = ({
  position,
  top,
  marginInlineStart,
  height,
  disabled = false,
  ...args
}) => {
  const [show, toggle] = useState(false);

  const toggleDropdown = () => {
    if (!disabled) {
      toggle((show) => !show);
    }
  };

  return (
    <ScrollableContainer>
      <div style={{ position, top, marginInlineStart, height }}>
        <Dropdown {...args} disabled={disabled}>
          <Button
            color="primary"
            as="div"
            className="filter-button"
            onClick={toggleDropdown}
            disabled={disabled}
          >
            Click to {show ? "close" : "open"}
          </Button>
        </Dropdown>
      </div>
    </ScrollableContainer>
  );
};

export const ScrollableContainerTopStart = ScrollableTemplate.bind({});
export const ScrollableContainerBottomStart = ScrollableTemplate.bind({});
export const ScrollableContainerTopEnd = ScrollableTemplate.bind({});
export const ScrollableContainerBottomEnd = ScrollableTemplate.bind({});

ScrollableContainerTopStart.args = {
  list: smallDropdownList,
  position: "relative",
  top: "0",
  height: "fit-content",
  fixPlacement: true,
};

ScrollableContainerBottomStart.args = {
  list: smallDropdownList,
  position: "relative",
  top: "200px",
  height: "fit-content",
  fixPlacement: true,
};

ScrollableContainerTopEnd.args = {
  list: smallDropdownList,
  marginInlineStart: "auto",
  height: "fit-content",
  fixPlacement: true,
};

ScrollableContainerBottomEnd.args = {
  list: smallDropdownList,
  marginInlineStart: "auto",
  position: "relative",
  top: "200px",
  height: "fit-content",
  fixPlacement: true,
};

export const WithInfoOnListOpen = Template.bind({});

WithInfoOnListOpen.args = {
  list: smallDropdownList,
  onToggleList(isListOpen) {
    alert(isListOpen);
  },
};
