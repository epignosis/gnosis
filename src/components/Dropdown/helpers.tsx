import { DropdownItem } from "./types";

export const filterListByKeyword = (list: DropdownItem[], keyword: string): DropdownItem[] => {
  return list.reduce((arr, item): DropdownItem[] => {
    const { items, label, originalText } = item;

    if (items && items.length > 0) {
      const filteredItems = filterListByKeyword(items, keyword);
      if (filteredItems.length > 0) {
        arr.push({ ...item, items: [...filteredItems] });
      }

      return arr;
    }

    if (typeof label === "string") {
      if (label.toLowerCase().includes(keyword.toLowerCase())) {
        arr.push(item);
      }
    } else if (originalText) {
      if (originalText.toLowerCase().includes(keyword.toLowerCase())) {
        arr.push(item);
      }
    }

    return arr;
  }, [] as DropdownItem[]);
};

// Reference: https://gist.github.com/twxia/bb20843c495a49644be6ea3804c0d775
export const getScrollableParent = (node: ParentNode | null): ParentNode | null => {
  const isElement = node instanceof HTMLElement;
  const overflowY = isElement && window.getComputedStyle(node).overflowY;
  const isScrollable = overflowY !== "visible" && overflowY !== "hidden";

  if (!node) {
    return null;
  } else if (isScrollable && isElement && node.scrollHeight >= node.clientHeight) {
    return node;
  }

  return getScrollableParent(node.parentNode) || null;
};

export const getInlinePaddingStart = (level: number, isSearchable: boolean) => {
  const initialPadding = isSearchable ? 0.25 : 1;

  return initialPadding + level * 0.75;
};

export const buildDropdownMenu = (list: DropdownItem[]): DropdownItem[] => {
  const hasNestedItems = list.some((item) => item.items && item.items.length > 0);

  if (!hasNestedItems) return list;

  const filteredGroups = list.filter((groupItem) => groupItem.items && groupItem.items.length > 0);

  return filteredGroups.reduce<DropdownItem[]>((result, groupItem, groupIndex) => {
    const isLastGroup = groupIndex === filteredGroups.length - 1;
    const groupItems = groupItem.items || [];
    const newGroup = [...groupItems];

    // Add divider to the last item of the group (except for the last group)
    if (!isLastGroup && newGroup.length > 0) {
      const lastItemIndex = newGroup.length - 1;
      newGroup[lastItemIndex] = { ...newGroup[lastItemIndex], divider: true };
    }
    return result.concat(newGroup);
  }, []);
};
