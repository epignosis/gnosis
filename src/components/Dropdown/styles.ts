import { Theme, css } from "@emotion/react";

const getInlinePaddingStart = (level: number, isSearchable: boolean) => {
  const initialPadding = isSearchable ? 0.25 : 1;

  return initialPadding + level * 0.75;
};

export const DropdownContainer = (
  { dropdown }: Theme,
  { isSearchable, fullWidth }: { isSearchable: boolean; fullWidth: boolean },
) => css`
  position: relative;
  display: inline-block;

  .dropdown-wrapper {
    position: absolute;
    z-index: 100;
    border-radius: 5px;
    box-shadow: 0 3px 6px ${dropdown.boxShadowColor};
    background-color: ${dropdown.backgroundColor};
    padding: ${isSearchable ? "0.75rem" : "0"};
    width: ${fullWidth ? (isSearchable ? "auto" : "100%") : "auto"};
    max-width: ${fullWidth ? "50" : "19.5"}rem;

    .input-wrapper {
      min-width: 18rem;
    }

    &.bottom-start {
      inset-inline-start: 0;
      margin-top: 0.25rem;
    }

    &.bottom-end {
      inset-inline-end: 0;
      margin-top: 0.25rem;
    }

    &.top-start {
      bottom: 100%;
      inset-inline-start: 0;
      margin-bottom: 0.25rem;
    }

    &.top-end {
      bottom: 100%;
      inset-inline-end: 0;
      margin-bottom: 0.25rem;
    }

    &.end-top {
      inset-inline-start: 100%;
      top: 0;
      margin: 0;
    }
  }
`;

export const DropdownList = (
  { dropdown, scrollbar }: Theme,
  { fullWidth, isSearchable }: { fullWidth: boolean; isSearchable: boolean },
) => css`
  list-style: none;
  width: max-content;
  box-sizing: border-box;
  white-space: nowrap;
  user-select: none;
  margin: ${isSearchable ? "0.5rem 0 0" : "0"};
  padding: ${isSearchable ? "0" : "0.5rem 0"};
  max-height: 21rem;
  overflow-y: auto;
  width: ${fullWidth ? "100%" : "auto"};
  min-width: ${fullWidth ? "2rem" : "14rem"};
  border-radius: 5px;

  /* Works on Firefox */
  scrollbar-width: 5px;
  scrollbar-color: ${scrollbar.color} ${scrollbar.background};

  /* Works on Chrome, Edge, and Safari */
  ::-webkit-scrollbar {
    width: 5px;
  }

  ::-webkit-scrollbar-track {
    background: ${scrollbar.background};
    border-radius: 10px;
  }

  ::-webkit-scrollbar-thumb {
    background-color: ${scrollbar.color};
    border-radius: 10px;
  }

  .empty-state {
    color: ${dropdown.emptyStateColor};
    padding: 0 0.75rem;
    display: flex;
    align-items: center;
    height: 2rem;
  }
`;

export const DropdownTitle = ({
  level,
  isSearchable,
}: {
  level: number;
  isSearchable: boolean;
}) => css`
  padding-inline: ${getInlinePaddingStart(level, isSearchable)}rem 1rem;
  padding-block: 0.25rem;
`;

export const DropdownListItem = (
  { dropdown }: Theme,
  {
    isSearchable,
    level,
    isDisabled,
  }: { isSearchable: boolean; level: number; isDisabled: boolean },
) => css`
  padding-inline: ${getInlinePaddingStart(level, isSearchable)}rem 1rem;
  padding-block: 0.5rem;
  display: flex;
  align-items: center;
  max-height: 2rem;

  .tooltip-content-wrapper {
    display: flex;
    align-items: center;
  }

  margin-inline-end: ${isSearchable ? "0.5rem" : "0"};
  cursor: ${isDisabled ? "default" : "pointer"};
  color: ${isDisabled ? dropdown.disabledColor : dropdown.textColor};

  &.separator {
    border-bottom: 1px solid ${dropdown.borderBottomColor};
  }

  span {
    word-break: break-word;
    overflow: hidden;
    text-overflow: ellipsis;
    display: inline-block;
    -webkit-line-clamp: 1;
    -webkit-box-orient: vertical;
  }

  svg {
    width: 32px;
    margin-inline-end: 0.5rem;
  }

  &:hover {
    background-color: ${isDisabled ? "inherit" : dropdown.hoverBackgroundColor};
    border-radius: ${isSearchable ? "5px" : "0"};
  }
`;
