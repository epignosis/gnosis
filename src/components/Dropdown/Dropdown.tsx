import React, {
  FC,
  Fragment,
  useEffect,
  useLayoutEffect,
  useRef,
  useState,
  MouseEvent,
  KeyboardEvent,
} from "react";
import classNames from "classnames";
import { SerializedStyles } from "@emotion/react";
import { useClickAway } from "ahooks";
import Text from "../Text/Text";
import SearchInput from "../FormElements/Input/SearchInput";
import Tooltip from "../Tooltip/Tooltip";
import { DropdownContainer, DropdownList, DropdownListItem, DropdownTitle } from "./styles";
import { DropdownItem, DropdownProps, PlacementOptions } from "./types";
import { filterListByKeyword, getScrollableParent } from "./helpers";

const dropdownWrapperClasses = (placement: PlacementOptions): string =>
  classNames("dropdown-wrapper", {
    "bottom-start": placement === "bottom-start",
    "bottom-end": placement === "bottom-end",
    "top-start": placement === "top-start",
    "top-end": placement === "top-end",
    "end-top": placement === "end-top",
  });

const dropdownButtonClasses = (isListOpen: boolean): string =>
  classNames("dropdown-button", {
    "is-active": isListOpen,
  });

const dropdownItemClasses = (item: DropdownItem): string =>
  classNames({
    "dropdown-list-item": true,
    [`${item.value}`]: true,
    [`${item.className}`]: Boolean(item.className),
  });

const RTLMapping: Record<PlacementOptions, PlacementOptions> = {
  "bottom-start": "bottom-end",
  "bottom-end": "bottom-start",
  "top-start": "top-end",
  "top-end": "top-start",
  "end-top": "end-top",
};

const Dropdown: FC<DropdownProps> = ({
  list,
  children,
  placement = "bottom-start",
  onListItemSelect,
  isSearchable,
  textSize = "sm",
  fullWidth = false,
  prependContent,
  hover = false,
  fixPlacement = false,
  emptyStateText = "No match found",
  placeholderText = "Search",
  remainOpenOnSelect = false,
  onToggleList,
  disabled = false,
  ...rest
}) => {
  const [isListOpen, setIsListOpen] = useState(false);
  const [currentPlacement, setCurrentPlacement] = useState(placement);
  const [filteredList, setFilteredList] = useState<DropdownItem[]>(() => list);
  const shouldFocus = Boolean(isSearchable);

  const wrapperRef = useRef<HTMLDivElement | null>(null);
  const dropdownButtonRef = useRef<HTMLDivElement | null>(null);
  const dropdownWrapperRef = useRef<HTMLDivElement | null>(null);

  const dir = document.dir;
  const isRtl = dir === "rtl";
  let hoverTimeOut: NodeJS.Timeout;

  useClickAway(() => {
    setIsListOpen(false);
  }, wrapperRef);

  useEffect(() => {
    setFilteredList(list);
  }, [list]);

  useEffect(() => {
    if (!disabled && onToggleList) {
      onToggleList(isListOpen);
    }
  }, [isListOpen]);

  useEffect(() => {
    if (disabled) {
      setIsListOpen(false);
      setFilteredList(list);
    }
  }, [disabled, list]);

  const toggleList = (e: MouseEvent<HTMLDivElement>): void => {
    e.stopPropagation();
    if (disabled || hover) return;

    // We want to reset the dropdown list every time it opens
    if (!isListOpen) {
      setFilteredList(list);
    }
    setIsListOpen((prevState) => !prevState);
  };

  const fixDropdownPlacement = (): void => {
    if (!dropdownButtonRef.current || !dropdownWrapperRef.current || !wrapperRef.current) return;

    const dropdownRect = dropdownButtonRef.current.getBoundingClientRect();
    const scrollableParent = getScrollableParent(wrapperRef.current.parentNode);
    const dropdownMenuHeight = dropdownWrapperRef.current.getBoundingClientRect().height;
    const dropdownMenuWidth = dropdownWrapperRef.current.getBoundingClientRect().width;

    let vertical = "bottom";
    let horizontal = "start";

    // Each case works with the same logic, but for different placements
    // We check if there is enough space on the right or left and top or bottom
    // and then we decide where to place the dropdown list
    // The dropdown list will be placed on the opposite side if there is not enough space
    // If there is a scrollable parent, we calculate the space based on the parent's dimensions
    // If there is no scrollable parent, we calculate the space based on the window's dimensions
    switch (placement) {
      case "bottom-start": {
        const dropdownButtonBottom = dropdownRect.bottom;
        if (!scrollableParent) {
          const spaceOnTheRight = window.innerWidth - dropdownRect.right;
          const spaceBelow = window.innerHeight - dropdownButtonBottom;

          horizontal = spaceOnTheRight < dropdownMenuWidth ? "end" : "start";
          vertical = spaceBelow < dropdownMenuHeight ? "top" : "bottom";
        }

        if (scrollableParent instanceof HTMLElement) {
          const spaceOnTheRight = scrollableParent.clientWidth - dropdownRect.right;
          const spaceBelow =
            scrollableParent.clientHeight -
            (dropdownButtonBottom - scrollableParent.getBoundingClientRect().top);

          horizontal = spaceOnTheRight < dropdownMenuWidth ? "end" : "start";
          vertical = spaceBelow < dropdownMenuHeight ? "top" : "bottom";
        }
        break;
      }

      case "bottom-end": {
        const dropdownButtonBottom = dropdownRect.bottom;
        if (!scrollableParent) {
          const spaceOnTheLeft = dropdownRect.left;
          const spaceBelow = window.innerHeight - dropdownButtonBottom;

          horizontal = spaceOnTheLeft < dropdownMenuWidth ? "start" : "end";
          vertical = spaceBelow < dropdownMenuHeight ? "top" : "bottom";
        }

        if (scrollableParent instanceof HTMLElement) {
          const spaceOnTheLeft = dropdownRect.left;
          const spaceBelow =
            scrollableParent.clientHeight -
            (dropdownButtonBottom - scrollableParent.getBoundingClientRect().top);

          vertical = spaceBelow < dropdownMenuHeight ? "top" : "bottom";
          horizontal = spaceOnTheLeft < dropdownMenuWidth ? "start" : "end";
        }
        break;
      }

      case "top-start": {
        if (!scrollableParent) {
          const spaceAbove = dropdownRect.top;
          const spaceOnTheRight = window.innerWidth - dropdownRect.right;

          horizontal = spaceOnTheRight < dropdownMenuWidth ? "end" : "start";
          vertical = spaceAbove < dropdownMenuHeight ? "bottom" : "top";
        }
        if (scrollableParent instanceof HTMLElement) {
          const spaceAbove = dropdownRect.top - scrollableParent.getBoundingClientRect().top;
          const spaceOnTheRight = scrollableParent.clientWidth - dropdownRect.right;

          horizontal = spaceOnTheRight < dropdownMenuWidth ? "end" : "start";
          vertical = spaceAbove < dropdownMenuHeight ? "bottom" : "top";
        }
        break;
      }

      case "top-end": {
        if (!scrollableParent) {
          const spaceAbove = dropdownRect.top;
          const spaceOnTheLeft = dropdownRect.left;

          horizontal = spaceOnTheLeft < dropdownMenuWidth ? "start" : "end";
          vertical = spaceAbove < dropdownMenuHeight ? "bottom" : "top";
        }

        if (scrollableParent instanceof HTMLElement) {
          const spaceAbove = dropdownRect.top - scrollableParent.getBoundingClientRect().top;
          const spaceOnTheLeft = dropdownRect.left;

          horizontal = spaceOnTheLeft < dropdownMenuWidth ? "start" : "end";
          vertical = spaceAbove < dropdownMenuHeight ? "bottom" : "top";
        }
        break;
      }
    }

    let res = `${vertical}-${horizontal}` as PlacementOptions;

    if (isRtl) {
      res = RTLMapping[res];
    }

    setCurrentPlacement(res);
  };

  useLayoutEffect(() => {
    if (isListOpen) {
      fixPlacement && fixDropdownPlacement();
    }
  }, [isListOpen]);

  const handleOnClickListItem = (item: DropdownItem): void => {
    if (isListOpen && !disabled) {
      onListItemSelect && onListItemSelect(item);
      !remainOpenOnSelect && setIsListOpen(false);
    }
  };

  const handleOnKeyDownListItem = (e: KeyboardEvent<HTMLLIElement>, item: DropdownItem): void => {
    if (e.key === "Enter" || e.key === " ") {
      handleOnClickListItem(item);
    }
  };

  const renderItemsRecursively = (items: DropdownItem[], level = 0): JSX.Element[] => {
    return items.map((item, index) => {
      const { isDisabled = false } = item;
      if (item.items) {
        return (
          <Fragment key={`${index}-${item.value}`}>
            <li css={DropdownTitle({ level, isSearchable: Boolean(isSearchable) })}>
              {typeof item.label === "string" ? (
                <Text fontSize={textSize} weight="700" title={item.label}>
                  {item.label}
                </Text>
              ) : (
                item.label
              )}
            </li>
            {renderItemsRecursively(item.items, level + 1)}
          </Fragment>
        );
      }

      const content = (
        <>
          {item?.icon}
          <Text fontSize={textSize} title={typeof item.label === "string" ? item.label : ""}>
            {item.label}
          </Text>
        </>
      );

      const tooltipElement = (
        <Tooltip
          key={`${index}-${item.value}`}
          disabled={!item.tooltipContent}
          content={item.tooltipContent}
          parentProps={{ className: "tooltip-content-wrapper" }}
        >
          {content}
        </Tooltip>
      );

      return (
        <li
          className={dropdownItemClasses(item)}
          tabIndex={0}
          key={`item-${index}-${item.value}`}
          data-testid={item.id}
          onClick={(): void => handleOnClickListItem(item)}
          onKeyDown={(e): void => handleOnKeyDownListItem(e, item)}
          css={(theme): SerializedStyles =>
            DropdownListItem(theme, {
              isSearchable: Boolean(isSearchable),
              level,
              isDisabled,
            })
          }
        >
          {tooltipElement}
        </li>
      );
    });
  };

  const handleInputChanged = (keyword: string): void => {
    if (!keyword) setFilteredList(list);
    setFilteredList(filterListByKeyword(list, keyword));
  };

  const handleOnMouseOver = () => {
    if (!hover) return;
    if (hoverTimeOut) clearTimeout(hoverTimeOut);

    setIsListOpen(true);
  };

  const handleOnMouseLeave = () => {
    if (!hover) return;

    // A small debounce to prevent dropdown from closing.
    hoverTimeOut = setTimeout(() => {
      setIsListOpen(false);
      // Force the list to be reset when the dropdown is closed
      setFilteredList(list);
    }, 100);
  };

  const handleOnKeyDown = (e: KeyboardEvent<HTMLDivElement>): void => {
    if (e.key === "Enter" || e.key === "Space") {
      setIsListOpen((prevState) => !prevState);
    }
  };

  return (
    <div
      css={(theme): SerializedStyles =>
        DropdownContainer(theme, { isSearchable: Boolean(isSearchable), fullWidth })
      }
      className="dropdown"
      ref={wrapperRef}
      tabIndex={0}
      onKeyDown={handleOnKeyDown}
      onMouseOver={handleOnMouseOver}
      onMouseLeave={handleOnMouseLeave}
      {...rest}
    >
      <div
        className={dropdownButtonClasses(isListOpen)}
        ref={dropdownButtonRef}
        onClick={toggleList}
      >
        {children}
      </div>

      {isListOpen && (
        <div
          className={dropdownWrapperClasses(fixPlacement ? currentPlacement : placement)}
          ref={dropdownWrapperRef}
        >
          {prependContent}

          {isSearchable && (
            <SearchInput
              placeholder={placeholderText}
              onInputChanged={handleInputChanged}
              id="dropdown-search"
              delayBeforeSearch={300}
              autoFocus={shouldFocus}
              disabled={disabled}
            />
          )}

          <ul
            css={(theme): SerializedStyles =>
              DropdownList(theme, { fullWidth, isSearchable: Boolean(isSearchable) })
            }
            role="list"
            className="dropdown-list"
          >
            {filteredList?.length ? (
              renderItemsRecursively(filteredList)
            ) : (
              <li className="empty-state">
                <Text fontSize="xs" weight="400">
                  {emptyStateText}
                </Text>
              </li>
            )}
          </ul>
        </div>
      )}
    </div>
  );
};

export default Dropdown;
