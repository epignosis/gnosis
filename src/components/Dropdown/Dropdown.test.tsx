import React from "react";
import Button from "../Button/Button";
import Dropdown from "./Dropdown";
import { DropdownItem } from "./types";
import { fireEvent, render, screen } from "@test-utils/render";

describe("<Dropdown />", () => {
  window.HTMLElement.prototype.scrollIntoView = jest.fn();

  it("opens the dropdown and matches snapshot", () => {
    const { getByText, container } = render(
      <Dropdown list={mockList}>
        <Button color="primary">Toggle</Button>
      </Dropdown>,
    );

    fireEvent.click(getByText("Toggle"));

    expect(container).toMatchSnapshot();
  });

  it("calls onListItemSelect when an item is clicked", () => {
    const mockOnListItemSelect = jest.fn();
    const { getByText } = render(
      <Dropdown list={mockList} onListItemSelect={mockOnListItemSelect}>
        <Button color="primary">Toggle</Button>
      </Dropdown>,
    );

    fireEvent.click(getByText("Toggle"));
    fireEvent.click(getByText("Option 2"));

    expect(mockOnListItemSelect).toHaveBeenCalledWith({
      id: "option-2",
      label: "Option 2",
      value: "2",
    });
  });
});

it("disabled dropdown does not call onToggleList when clicked", () => {
  const mockOnToggleList = jest.fn();
  const { getByText } = render(
    <Dropdown list={mockList} onToggleList={mockOnToggleList} disabled>
      <Button disabled color="primary">
        Toggle
      </Button>
    </Dropdown>,
  );

  const button = screen.getByRole("button");

  expect(button).toBeDisabled();

  fireEvent.click(getByText("Toggle"));

  expect(mockOnToggleList).not.toHaveBeenCalled();
});

const mockList: DropdownItem[] = [
  {
    id: "category-1",
    value: "category-1",
    label: "Category 1",
    items: [
      { id: "option-1", label: "Option 1", value: "1" },
      {
        id: "category-2",
        label: "Category 2",
        items: [
          { id: "option-2", label: "Option 2", value: "2" },
          { id: "option-3", label: "Option 3", value: "3" },
        ],
      },
    ],
  },
  {
    id: "category-15",
    value: "category-15",
    label: "Category 15",
    items: [
      { id: "option-15", label: "Option 15", value: "15" },
      { id: "option-16", label: "Option 16", value: "16" },
    ],
  },
  { id: "option-5", label: "Option 5", value: "5" },
];
