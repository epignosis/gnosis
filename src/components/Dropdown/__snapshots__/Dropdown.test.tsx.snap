// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> opens the dropdown and matches snapshot 1`] = `
<div>
  <div
    class="dropdown css-s6eb27-DropdownContainer-DropdownContainer"
    tabindex="0"
  >
    <div
      class="dropdown-button is-active"
    >
      <button
        class="solid css-z4fqld-baseButton-baseButton-solidButton-solidButton-outlineButton-outlineButton-ghostButton-ghostButton-linkButton-linkButton-iconBeforeButton-iconBeforeButton-iconAfterButton-iconAfterButton-btnContainer-btnContainer"
      >
        <span
          class="btn-text"
        >
          Toggle
        </span>
      </button>
    </div>
    <div
      class="dropdown-wrapper bottom-start"
    >
      <ul
        class="dropdown-list css-1an3w4m-DropdownList-DropdownList"
        role="list"
      >
        <li
          class="css-rndanj-DropdownTitle-DropdownTitle"
        >
          <span
            class="css-mproha-textContainer-textContainer"
            data-testid="text-component"
            title="Category 1"
          >
            Category 1
          </span>
        </li>
        <li
          class="1 dropdown-list-item css-12omqks-DropdownListItem-DropdownListItem"
          tabindex="0"
        >
          <div
            aria-expanded="false"
            class="tooltip-content-wrapper"
            role="tooltip"
          >
            <span
              class="css-1by32j1-textContainer-textContainer"
              data-testid="text-component"
              title="Option 1"
            >
              Option 1
            </span>
          </div>
        </li>
        <li
          class="css-1mnx8ve-DropdownTitle-DropdownTitle"
        >
          <span
            class="css-mproha-textContainer-textContainer"
            data-testid="text-component"
            title="Category 2"
          >
            Category 2
          </span>
        </li>
        <li
          class="2 dropdown-list-item css-mrayhu-DropdownListItem-DropdownListItem"
          tabindex="0"
        >
          <div
            aria-expanded="false"
            class="tooltip-content-wrapper"
            role="tooltip"
          >
            <span
              class="css-1by32j1-textContainer-textContainer"
              data-testid="text-component"
              title="Option 2"
            >
              Option 2
            </span>
          </div>
        </li>
        <li
          class="3 dropdown-list-item css-mrayhu-DropdownListItem-DropdownListItem"
          tabindex="0"
        >
          <div
            aria-expanded="false"
            class="tooltip-content-wrapper"
            role="tooltip"
          >
            <span
              class="css-1by32j1-textContainer-textContainer"
              data-testid="text-component"
              title="Option 3"
            >
              Option 3
            </span>
          </div>
        </li>
        <li
          class="css-rndanj-DropdownTitle-DropdownTitle"
        >
          <span
            class="css-mproha-textContainer-textContainer"
            data-testid="text-component"
            title="Category 15"
          >
            Category 15
          </span>
        </li>
        <li
          class="15 dropdown-list-item css-12omqks-DropdownListItem-DropdownListItem"
          tabindex="0"
        >
          <div
            aria-expanded="false"
            class="tooltip-content-wrapper"
            role="tooltip"
          >
            <span
              class="css-1by32j1-textContainer-textContainer"
              data-testid="text-component"
              title="Option 15"
            >
              Option 15
            </span>
          </div>
        </li>
        <li
          class="16 dropdown-list-item css-12omqks-DropdownListItem-DropdownListItem"
          tabindex="0"
        >
          <div
            aria-expanded="false"
            class="tooltip-content-wrapper"
            role="tooltip"
          >
            <span
              class="css-1by32j1-textContainer-textContainer"
              data-testid="text-component"
              title="Option 16"
            >
              Option 16
            </span>
          </div>
        </li>
        <li
          class="5 dropdown-list-item css-1hv1td9-DropdownListItem-DropdownListItem"
          tabindex="0"
        >
          <div
            aria-expanded="false"
            class="tooltip-content-wrapper"
            role="tooltip"
          >
            <span
              class="css-1by32j1-textContainer-textContainer"
              data-testid="text-component"
              title="Option 5"
            >
              Option 5
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
